File: redhawk/codegen/jinja/java/ports/templates/message.java
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
//% set classname = portgenerator.className()
package ${package};

import java.util.HashMap;
import java.util.Map;
import org.ossie.component.QueryableUsesPort;
import java.util.ArrayList;
import java.util.List;
import CF.PropertiesHelper;
import CF.DataType;
import org.omg.CORBA.ORB;
import org.omg.CORBA.Any;
import ${component.package}.${component.baseclass.name}.*;
import org.ossie.events.*;

/**
 * @generated
 */
public class ${classname} extends MessageSupplierPort {
    public ${classname}(String portName) {
        super(portName);
    }
/*{% for prop in component.messages %}*/

    public void sendMessage(final ${prop.javatype} message) {
        final List<DataType> outProps = new ArrayList<DataType>();
        final List<DataType> propStruct = new ArrayList<DataType>();
        Any propStruct_any = ORB.init().create_any();
/*{%   for field in prop.fields %}*/
        propStruct.add(new DataType("${field.identifier}", message.${field.javaname}.toAny()));
/*{%   endfor %}*/
        PropertiesHelper.insert(propStruct_any, propStruct.toArray(new DataType[propStruct.size()]));
        outProps.add(new DataType("${prop.identifier}", propStruct_any));
        Any outProps_any = ORB.init().create_any();
        PropertiesHelper.insert(outProps_any, outProps.toArray(new DataType[outProps.size()]));
        this.push(outProps_any);
    }
    
    public void sendMessages(final ArrayList<${prop.javatype}> messages) {
        final List<DataType> outProps = new ArrayList<DataType>();
        for (${prop.javatype} message : messages) {
            final List<DataType> propStruct = new ArrayList<DataType>();
            Any propStruct_any = ORB.init().create_any();
/*{%   for field in prop.fields %}*/
            propStruct.add(new DataType("${field.identifier}", message.${field.javaname}.toAny()));
/*{%   endfor %}*/
            PropertiesHelper.insert(propStruct_any, propStruct.toArray(new DataType[propStruct.size()]));
            outProps.add(new DataType("${prop.identifier}", propStruct_any));
        }
        Any outProps_any = ORB.init().create_any();
        PropertiesHelper.insert(outProps_any, outProps.toArray(new DataType[outProps.size()]));
        this.push(outProps_any);
    }
/*{% endfor %}*/
}

File: redhawk/codegen/jinja/java/ports/message.py
#
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of REDHAWK core.
#
# REDHAWK core is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.
#

import jinja2

from redhawk.codegen.lang import java
from redhawk.codegen.jinja.java import JavaTemplate
from redhawk.codegen.jinja.ports import PortFactory

from generator import JavaPortGenerator, BuiltinJavaPort

if not '__package__' in locals():
    # Python 2.4 compatibility
    __package__ = __name__.rsplit('.', 1)[0]

class MessagePortFactory(PortFactory):
    REPID = 'IDL:ExtendedEvent/MessageEvent:1.0'

    def match(cls, port):
        return (port.repid() == cls.REPID)

    def generator(cls, port):
        if port.isProvides():
            return MessageConsumerPortGenerator(port)
        else:
            return MessageSupplierPortGenerator(port)

class MessageConsumerPortGenerator(BuiltinJavaPort):
    def __init__(self, port):
        BuiltinJavaPort.__init__(self, 'org.ossie.events.MessageConsumerPort', port)

    def _ctorArgs(self, name):
        return (java.stringLiteral(name),)

class MessageSupplierPortGenerator(JavaPortGenerator):
    def loader(self):
        return jinja2.PackageLoader(__package__)

    def _implementation(self):
        return JavaTemplate('message.java')

    def _ctorArgs(self, name):
        return (java.stringLiteral(name),)
